<p align="center">
    <img src="https://raw.githubusercontent.com/mmtechslv/PhyloMAF/master/logo.png?token=AI5GEEUPZJWXRVI233DEVUTA3D3HI" width="1280" height="640">
</p>
<p align="center">
    PhyloMAF - Phylogenetic Microbiome Analysis Framework
</p>
<p align="center">
    <img src="https://travis-ci.com/mmtechslv/PhyloMAF.svg?token=sBXDx4EANVBbNzh5fzox&branch=master" alt="Build Status">
</p>
<p>
    PhyloMAF is a novel comprehensive microbiome data analysis tool based on Python programming language. With memory efficient and extensible design, PhyloMAF have wide range of applications including but not limited to: post OTU picking microbiome data analysis, microbiome data meta-analysis, taxonomy based reference phylogenetic tree pruning and reconstruction, cross database data validation, primer design by taxonomic ranks, heterogeneous data retrieval from different databases including remote mega-databases like NCBI or Ensembl.
</p>

NOTE: Project will be uploaded during article publication process.

