<p align="center">
    <img src="https://github.com/mmtechslv/PhyloMAF/blob/c1132dae7a778811a6abd9bafd329b6067916396/logo.png">
</p>
<p align="center">
    PhyloMAF - Phylogenetic Microbiome Analysis Framework
</p>
<p align="center">
    <img src="https://travis-ci.com/mmtechslv/PhyloMAF.svg?token=sBXDx4EANVBbNzh5fzox&branch=master" alt="Build Status">
    <a href='https://phylomaf.readthedocs.io/en/latest/?badge=latest'>
        <img src='https://readthedocs.org/projects/phylomaf/badge/?version=latest' alt='Documentation Status' />
    </a>
    <a href="https://zenodo.org/badge/latestdoi/197987819">
        <img src="https://zenodo.org/badge/197987819.svg" alt="DOI">
    </a>
</p>
<p>
    PhyloMAF is a novel comprehensive microbiome data analysis tool based on Python programming language. With memory efficient and extensible design, PhyloMAF have wide range of applications including but not limited to: post OTU picking microbiome data analysis, microbiome data meta-analysis, taxonomy based reference phylogenetic tree pruning and reconstruction, cross database data validation, primer design by taxonomic ranks, heterogeneous data retrieval from different databases including remote mega-databases like NCBI or Ensembl.
</p>


